pycqed/instrument_drivers/meta_instrument/device.py

```diff
@@ -0,0 +1,991 @@
+        cp = CalibrationPoints.single_qubit(qbr.name, cal_states,
+                                            n_per_state=n_cal_points_per_state)
+
+        if prep_params is None:
+            prep_params = self.get_prep_params([qbc, qbt])
+
+        operation_dict = self.get_operation_dict()
+
+        sequences, hard_sweep_points, soft_sweep_points = \
+            fsqs.chevron_seqs(
+                qbc_name=qbc.name, qbt_name=qbt.name,
+                hard_sweep_dict=hard_sweep_params,
+                soft_sweep_dict=soft_sweep_params,
+                operation_dict=operation_dict,
+                cz_pulse_name=cz_pulse_name,
+                num_cz_gates=num_cz_gates,
+                cal_points=cp, upload=False, prep_params=prep_params)
+
+        if return_seq:
+            return sequences
+
+        hard_sweep_func = awg_swf.SegmentHardSweep(
+            sequence=sequences[0], upload=upload,
+            parameter_name=list(hard_sweep_params)[0],
+            unit=list(hard_sweep_params.values())[0]['unit'])
+        MC.set_sweep_function(hard_sweep_func)
+        MC.set_sweep_points(hard_sweep_points)
+
+        # sweep over flux pulse amplitude of qbc
+        channels_to_upload = [qbc.flux_pulse_channel()]
+        MC.set_sweep_function_2D(awg_swf.SegmentSoftSweep(
+            hard_sweep_func, sequences,
+            list(soft_sweep_params)[0], list(soft_sweep_params.values())[0]['unit'],
+            channels_to_upload=channels_to_upload))
+        MC.set_sweep_points_2D(soft_sweep_points)
+        MC.set_detector_function(qbr.int_avg_classif_det if classified
+                                 else qbr.int_avg_det)
+        if exp_metadata is None:
+            exp_metadata = {}
+        exp_metadata.update({'preparation_params': prep_params,
+                             'cal_points': repr(cp),
+                             'rotate': len(cal_states) != 0,
+                             'data_to_fit': {qbr.name: 'pe'},
+                             'hard_sweep_params': hard_sweep_params,
+                             'soft_sweep_params': soft_sweep_params})
+        MC.run_2D(name=label, exp_metadata=exp_metadata)
+
+        if analyze:
+            tda.MultiQubit_TimeDomain_Analysis(qb_names=[qbr.name],
+                                               options_dict={'TwoD': True})
+
+    def measure_cphase(self, qbc, qbt, soft_sweep_params, cz_pulse_name=None, prep_params=None,
+                       label=None, cal_states='auto', n_cal_points_per_state=1,
+                       num_cz_gates=1, hard_sweep_params=None, exp_metadata=None,
+                       analyze=True, upload=True, for_ef=True, **kw):
+
+        qbc_name = qbc.name
+        qbt_name = qbt.name
+
+        if [qbc_name, qbt_name] not in self.connectivity_graph() and [qbt_name,
+                                                                      qbc_name] not in self.connectivity_graph():
+            raise ValueError('Qubits are not connected!')
```

Comment:
I would suggest a wrapper to the device class to do this, it will be useful also at some other places and we might be able to build upon it when using swaps.:
``` 
def check_connection(self, qubit_a, qubit_b, connectivity_graph=None, raise_exception=True):

    if connectivity_graph is None:
        connectivity_graph = self.connectivity_graph()
    # convert qubit object to name if necessary
    if not isintance(qubit_a, str):
        qubit_a = qubit_a.name
    if not isinstance(qubit_b, str):
        qubit_b = qubit_b.name
    if [qubit_a, qubit_b] not in self.connectivity_graph() and 
        [qubit_b, qubit_a] not in self.connectivity_graph():
        if raise_exception:
            raise ValueError(f'Qubits {[qubit_a, qubit_b]}  are not connected!')
        else:
            log.warning('Qubits are not connected!')
            # TODO: implement what we want in case of swap (e.g. determine shortest path of swaps)
            
```

Reply:
just misses a doc string ;)

Reply:
Nice! Now implemented in the code!